mass_mailing_partner/models/mail_mass_mailing_contact.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2015 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# Copyright 2015 Antonio Espinosa <antonio.espinosa@tecnativa.com>
+# Copyright 2015 Javier Iniesta <javieria@antiun.com>
+# Copyright 2017 David Vidal <david.vidal@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class MailMassMailingContact(models.Model):
+    _inherit = 'mail.mass_mailing.contact'
+
+    partner_id = fields.Many2one(comodel_name='res.partner', string="Partner",
+                                 domain=[('email', '!=', False)])
+
+    @api.constrains('partner_id', 'list_ids')
+    def _check_partner_id_lists_ids(self):
+        for contact in self:
+            if contact.partner_id:
+                other_contact = self.search([
+                    ('partner_id', '=', contact.partner_id.id),
+                    ('id', '!=', contact.id)
+                ])
+                if contact.list_ids & other_contact.mapped('list_ids'):
+                    raise ValidationError(_("Partner already exists in one of "
+                                            "these mailing lists"))
+
+    @api.onchange('partner_id')
+    def _onchange_partner(self):
+        if self.partner_id:
+            self.name = self.partner_id.name
+            self.email = self.partner_id.email
+            if self.partner_id.title:
+                self.title_id = self.partner_id.title.id
```

Comment:
When doing assignment with this kind of notation, it's safer to do instead:

```py
self.title_id = self.partner_id.title
```

Because the ORM will check both records belong to the same model, and the rest will work equally.